### PR TITLE
Disable memory sharing on model parameters in ddp-spawn 

### LINKED ISSUE #17399: Cannot replicate training results with seed_everything and `strategy=ddp_spawn`
### Bug description

I noticed that `strategy=ddp_spawn` can not give deterministic results. If I change `strategy=ddp_spawn` to `strategy=ddp_fork`, then training results are deterministic. I use `pytorch_lightning==1.9.5` and `torch==1.11.0`

Below is the code that can reproduce the bug.

```py
import copy
import os
from typing import Dict, List

import numpy as np
import pytorch_lightning as pl
import torch
import torch.distributed as dist
import torch.nn.functional as F
from torch import nn
from torch.utils.data import DataLoader, random_split
from torchvision import transforms
from torchvision.datasets import MNIST


class LitAutoEncoder(pl.LightningModule):
    def __init__(self):
        super().__init__()
        self.encoder = nn.Sequential(nn.Linear(28 * 28, 128), nn.ReLU(), nn.Linear(128, 3))
        self.decoder = nn.Sequential(nn.Linear(3, 128), nn.ReLU(), nn.Linear(128, 28 * 28))

    def forward(self, x):
        # in lightning, forward defines the prediction/inference actions
        embedding = self.encoder(x)
        return embedding

    def training_step(self, batch, batch_idx):
        # training_step defines the train loop. It is independent of forward
        x, y = batch
        x = x.view(x.size(0), -1)
        z = self.encoder(x)
        x_hat = self.decoder(z)
        loss = F.mse_loss(x_hat, x)
        self.log("train_loss", loss)
        return loss

    def configure_optimizers(self):
        optimizer = torch.optim.Adam(self.parameters(), lr=1e-3)
        return optimizer

dataset = MNIST(os.getcwd(), download=True, transform=transforms.ToTensor())
train, val = random_split(dataset, [55000, 5000])

def compare(strategy):
    pl.seed_everything(5)
    torch.set_num_threads(1)
    autoencoder = LitAutoEncoder()
    trainer = pl.Trainer(
        max_epochs=1,
        accelerator="cpu",
        devices=2,
        enable_progress_bar=False,
        enable_model_summary=False,
        deterministic=True,
        strategy=strategy,
    )
    trainer.fit(
        autoencoder, DataLoader(train, batch_size=500), DataLoader(val, batch_size=500)
    )

    pl.seed_everything(5)
    torch.set_num_threads(1)
    copy_autoencoder = LitAutoEncoder()
    copy_trainer = pl.Trainer(
        max_epochs=1,
        accelerator="cpu",
        devices=2,
        enable_progress_bar=False,
        enable_model_summary=False,
        deterministic=True,
        strategy=strategy,
    )
    copy_trainer.fit(
        copy_autoencoder, DataLoader(train, batch_size=500), DataLoader(val, batch_size=500)
    )

    def assert_state_dict_equal(state_dict_a: dict, state_dict_b: dict):
        np.testing.assert_equal(list(state_dict_a.keys()), list(state_dict_b.keys()))
        for key in state_dict_a:
            np.testing.assert_array_equal(state_dict_a[key], state_dict_b[key])

    assert_state_dict_equal(
        copy.deepcopy(autoencoder).state_dict(),
        copy.deepcopy(copy_autoencoder).state_dict(),
    )

if __name__ == '__main__':
    strategy = "ddp_spawn"
    compare(strategy)

```

### What version are you seeing the problem on?

1.9.x

### How to reproduce the bug

_No response_

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 1.5.0):
#- Lightning App Version (e.g., 0.5.2):
#- PyTorch Version (e.g., 2.0):
#- Python version (e.g., 3.9):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
#- Running environment of LightningApp (e.g. local, cloud):
```

</details>

### More info

_No response_

cc @awaelchli @justusschock

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #17399

The `torch.multiprocessing.spawn` launcher (strategy="ddp_spawn") by default enables memory sharing for all tensors passed through the spawning function, including tensors in modules. This means that the underlying storage of these tensors is shared across all processes, and anyone can read or write to them. This can lead to inconsistencies, as demonstrated in the repro example in the linked issue. This PR disables that by cloning the weights in each process, detaching it from shared memory.

Note: 
This only applies when running on CPU. If running on GPU, memory won't be shared.
Hopefully, this will also help to avoid flakiness in tests.

cc @borda @carmocca @justusschock @awaelchli

### CODE REVIEW

File: src/lightning/fabric/strategies/launchers/multiprocessing.py [hunk 1]
```diff
@@ -187,3 +192,14 @@ def _check_bad_cuda_fork() -> None:
     if _IS_INTERACTIVE:
         message += " You will have to restart the Python kernel."
     raise RuntimeError(message)
+
+
+def _disable_module_memory_sharing(data: Any) -> None:
```
Comment (awaelchli) [inline src/lightning/fabric/strategies/launchers/multiprocessing.py:197]: TODO: try the alternative of deepcopy

File: src/lightning/fabric/strategies/launchers/xla.py [hunk 2]
```diff
@@ -100,12 +100,9 @@ def _wrapping_function(
     ) -> None:
         import torch_xla.core.xla_model as xm
 
+        # `get_xla_supported_devices` in the spawned process returns the logical devices (2 for v2/v3 and 1 for v4)
         if _using_pjrt() and len(xm.get_xla_supported_devices()) > 1:
-            # `get_xla_supported_devices` in the spawned process returns the logical devices (2 for v2/v3 and 1 for v4)
-            # so when there's more than one (multithreading), objects need to be deep-copied
-            import copy
-
-            function, args, kwargs = copy.deepcopy((function, args, kwargs))
```
Comment (carmocca) [inline src/lightning/fabric/strategies/launchers/xla.py:108]: @Liyang90 made the deepcopy in a single call to "maintain the reference links between these objects" (https://github.com/Lightning-AI/lightning/pull/17352#discussion_r1166274656). This PR changes this as `apply_to_collection` will not do this. I'll add the TPU label to the title so TPU tests run.

**Review by carmocca — COMMENTED** [review]
For the shared parameters, we could do what XLA already does: https://github.com/Lightning-AI/lightning/blob/master/src/lightning/pytorch/strategies/xla.py#L123-L126

File: src/lightning/pytorch/CHANGELOG.md [hunk 3]
```diff
@@ -200,6 +200,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Fixed an issue that would prevent the user to set the multiprocessing start method after importing lightning ([#18177](https://github.com/Lightning-AI/lightning/pull/18177))
 
 
+- Fixed model parameters getting shared between processes when running with `strategy="ddp_spawn` and `accelerator="cpu"` ([#18238](https://github.com/Lightning-AI/lightning/pull/18238))
```
Comment (awaelchli) [inline src/lightning/pytorch/CHANGELOG.md:203]:
```suggestion
- Fixed model parameters getting shared between processes when running with `strategy="ddp_spawn"` and `accelerator="cpu"` ([#18238](https://github.com/Lightning-AI/lightning/pull/18238))
```

File: src/lightning/fabric/CHANGELOG.md [hunk 4]
```diff
@@ -195,6 +195,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Fixed an issue with `Fabric.all_reduce()` not performing an inplace operation for all backends consistently ([#18235](https://github.com/Lightning-AI/lightning/pull/18235))
 
 
+- Fixed model parameters getting shared between processes when running with `strategy="ddp_spawn` and `accelerator="cpu"` ([#18238](https://github.com/Lightning-AI/lightning/pull/18238))
```
Comment (awaelchli) [inline src/lightning/fabric/CHANGELOG.md:198]:
```suggestion
- Fixed model parameters getting shared between processes when running with `strategy="ddp_spawn"` and `accelerator="cpu"` ([#18238](https://github.com/Lightning-AI/lightning/pull/18238))
```

File: src/lightning/fabric/strategies/launchers/xla.py (hunk 2 above)
Comment (awaelchli) [reply src/lightning/fabric/strategies/launchers/xla.py]: XLA uses threading here, so the purpose is different. I reverted the change. For the fix I'm doing here, deep-copying is not sufficient.

File: src/lightning/fabric/strategies/launchers/multiprocessing.py [hunk 5]
```diff
@@ -190,3 +198,21 @@ def _check_bad_cuda_fork() -> None:
...
+    """Disables memory sharing on parameters and buffers of `nn.Module`s contained in the given collection.
+
+    Note: This is only required when running on CPU.
+    """
+    # PyTorch enables memory sharing automatically on all tensors that are passed through `mp.spawn`.
+    # For model weights and buffers, this is undesired and can lead to race conditions between processes.
+    # Hence, we copy the tensors in the entire module to ensure it doesn't share memory with other processes.
+
+    @torch.no_grad()
+    def _unshare(module: Module) -> Module:
```
Comment (carmocca) [inline src/lightning/fabric/strategies/launchers/multiprocessing.py:213]: nit: local functions don't need to be protected

File: src/lightning/fabric/strategies/launchers/multiprocessing.py [hunk 6]
```diff
@@ -190,3 +198,21 @@ def _check_bad_cuda_fork() -> None:
...
+        for tensor in itertools.chain(module.parameters(), module.buffers()):
+            tensor.data = tensor.data.clone()
+        return module
+
+    return apply_to_collection(data, function=_unshare, dtype=Module)
```
Comment (carmocca) [inline src/lightning/fabric/strategies/launchers/multiprocessing.py:218]:
```suggestion
    return apply_to_collection(data, function=unshare, dtype=Module)
```

File: src/lightning/fabric/CHANGELOG.md [hunk 7]
```diff
@@ -198,6 +198,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Fixed an issue with `Fabric.all_reduce()` not performing an inplace operation for all backends consistently ([#18235](https://github.com/Lightning-AI/lightning/pull/18235))
 
 
+- Fixed model parameters getting shared between processes when running with `strategy="ddp_spawn"` and `accelerator="cpu"` ([#18238](https://github.com/Lightning-AI/lightning/pull/18238))
```
Comment (carmocca) [inline src/lightning/fabric/CHANGELOG.md:201]: This will come with a memory increase. Mention it explicitly?

**Review by carmocca — APPROVED** [review]

File: src/lightning/fabric/strategies/launchers/multiprocessing.py (hunk 5 above)
Comment (awaelchli) [reply src/lightning/fabric/strategies/launchers/multiprocessing.py:213]:
```suggestion
    def unshare(module: Module) -> Module:
```

**Review by Borda — APPROVED** [review]
